source/endpoints_regex.c

```diff
@@ -0,0 +1,312 @@
+                break;
+            case '+':
+                symbol.type = AWS_ENDPOINTS_REGEX_SYMBOL_PLUS;
+                break;
+            case '\\':
+                switch (regex_pattern.ptr[0]) {
+                    /* Predefined patterns */
+                    case 'd':
+                        symbol.type = AWS_ENDPOINTS_REGEX_SYMBOL_DIGIT;
+                        break;
+                    case 'w':
+                        symbol.type = AWS_ENDPOINTS_REGEX_SYMBOL_ALPHA;
+                        break;
+
+                    /* Escaped chars, ex. * or + */
+                    default:
+                        symbol.type = AWS_ENDPOINTS_REGEX_SYMBOL_CHAR;
+                        symbol.info.ch = regex_pattern.ptr[0];
+                        break;
+                }
+                aws_byte_cursor_advance(&regex_pattern, 1);
+                break;
+            case '(': {
+                struct aws_byte_cursor group = {0};
+                if (!aws_byte_cursor_next_split(&regex_pattern, ')', &group)) {
+                    AWS_LOGF_ERROR(
+                        AWS_LS_SDKUTILS_ENDPOINTS_REGEX, "Invalid regex pattern. Missing closing parenthesis.");
+                    aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+                    goto on_error;
+                }
+
+                aws_byte_cursor_advance(&regex_pattern, group.len);
+                if (regex_pattern.len == 0 || regex_pattern.ptr[0] != ')') {
+                    AWS_LOGF_ERROR(
+                        AWS_LS_SDKUTILS_ENDPOINTS_REGEX, "Invalid regex pattern. Missing closing parenthesis.");
+                    aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+                    goto on_error;
+                }
+                aws_byte_cursor_advance(&regex_pattern, 1);
+
+                if (group.len == 0) {
+                    AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_REGEX, "Invalid regex pattern. Empty group.");
+                    aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+                    goto on_error;
+                }
+
+                /* Verify that group is only used for alternation. */
+                for (size_t i = 0; i < group.len; ++i) {
+                    if (!aws_isalnum(group.ptr[i]) && group.ptr[i] != '|') {
+                        AWS_LOGF_ERROR(
+                            AWS_LS_SDKUTILS_ENDPOINTS_REGEX,
+                            "Unsupported regex pattern. Only alternation groups are supported.");
+                        aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_UNSUPPORTED_REGEX);
+                        goto on_error;
+                    }
+                }
+
+                symbol.type = AWS_ENDPOINTS_REGEX_SYMBOL_ALTERNATION_GROUP;
+                symbol.info.alternation = aws_string_new_from_cursor(allocator, &group);
+                break;
+            }
+
+            default: {
+                if (!aws_isalnum(ch)) {
+                    AWS_LOGF_ERROR(
```

Comment:
Trivial/Debatable: Printing the index would be helpful in error messages, but I suppose we are not expecting these to fail often.

Reply:
we dont keep track of the index and i would prefer to keep it simpler. 
this branch would only fallthrough when none of the existing chars match known symbols. And arguably if we dont support a symbol, ex. ?, it does not matter where is appears as none of the occurences of that symbol will be supported